### PR TITLE
KEYCLOAK-10603 adjust user-role and group-role assignments

### PR DESCRIPTION
**Pull request by danielFesenmeyer**
This PR adjusts user-role and group-role assignments as suggested in issue KEYCLOAK-10603: It allows assignments of roles which are already indirectly assigned.

When you are comfortable with the changes, I would adjust client- and client-scope role-mappings accordingly. I think it makes sense to have a consistent behavior for all sorts of role-mappings.

Basically, I made the following changes:
- extend RoleMapperModel with method hasDirectRole(RoleModel), which only checks for direct assignment in contrast to the existing method hasRole(RoleModel)
- use the new hasDirectRole method to check whether a role is in the "available" list and whether it can be assigned (previously, the hasRole method was used for this purpose)

Can someone please review my suggested changes and provide feedback? Thanks in advance :)

### CODE REVIEW

**Comment by stianst** [comment]
@mposolda @pedroigor can one of you take a look at this please?

File: server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java
Comment (pedroigor) [inline server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java:72]: Would be better to have this method as default?

Comment (pedroigor) [inline server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java:72]: Would be better to have this method as default?

**Review by pedroigor — COMMENTED** [review]
@danielFesenmeyer The fix makes and sense.

File: services/src/main/java/org/keycloak/services/resources/admin/ClientRoleMappingsResource.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/ClientRoleMappingsResource.java:137]: UX-wise admins should now see the client role available even though it is effectively granted, right?

Wondering if makes sense to update the hint for available roles to state that these roles are not directly granted to the user even though they might be shown as effective.

File: services/src/main/java/org/keycloak/services/resources/admin/RoleMapperResource.java
Comment (pedroigor) [inline services/src/main/java/org/keycloak/services/resources/admin/RoleMapperResource.java:203]: See the comment I did for client roles and if makes sense to change hint for available realm roles.

File: server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java
Comment (danielFesenmeyer) [reply server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java:72]: @pedroigor: What do you mean with having this method as default? Having a default implementation which delegates to hasRole(Role)?

File: services/src/main/java/org/keycloak/services/resources/admin/ClientRoleMappingsResource.java
Comment (danielFesenmeyer) [reply services/src/main/java/org/keycloak/services/resources/admin/ClientRoleMappingsResource.java:137]: @pedroigor Exactly, all roles which are not directly granted will be shown as available. Currently, for client roles, this hint is shown: "Assignable roles from this client". We can add a hint like "Contains effectively assigned roles which are not directly assigned." But we could only do this for English (and German), because we don't speak the other supported languages :(

Comment (pedroigor) [reply services/src/main/java/org/keycloak/services/resources/admin/ClientRoleMappingsResource.java:137]: Works for me.

File: server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java
Comment (pedroigor) [reply server-spi/src/main/java/org/keycloak/models/RoleMapperModel.java:72]: Yes, but perhaps using `getRoleMappingsStream`.

**Comment by danielFesenmeyer** [comment]
Hi @pedroigor,

I have integrated your suggestions into the PR:

- Method hasDirectRole now has a default implementation delegating to getRoleMappingsStream.
- updated the UI hints for available roles with the info  that they contain effectively assigned roles which are not directly assigned.

Currently the PR only considers user-role and group-role assignments, but the same issue applies to client- and client-scope role-mappings. I would also adjust this within this PR - or should I create a separate one?

**Comment by danielFesenmeyer** [comment]
Pushed a small update with a bugfix for failing integration test LDAPRoleMappingsNoImportTest.

**Review by pedroigor — DISMISSED** [review]

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/UserTest.java
Comment (pedroigor) [inline testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/UserTest.java:2185]: Don't you think that `realm-child` here is a bit confusing and susceptible to mistakes? Admins now see that role as available, but it was granted by the composite. Although indirectly.

I'm not 100% if we should deal with composite roles differently in this case. So that they are not listed as available.

**Comment by pedroigor** [comment]
@danielFesenmeyer The PR is smaller now, so perhaps we can include scope mapping too. From a reviewer's perspective, the pattern should be the same.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/UserTest.java
Comment (danielFesenmeyer) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/admin/UserTest.java:2185]: @pedroigor I agree that this might be a bit confusing, at least for users who are used to the old behaviour. But otherwise it would not be possible to assign "realm-child" via the UI (at least without bigger changes to the UI).
Furthermore, there are other parts of the UI, where it is possible to assign indirectly assigned roles which are not directly assigned. There, similar mistakes might happen:

- assigning default roles to a realm
- assigning a child to a composite role

**Comment by danielFesenmeyer** [comment]
@pedroigor, @mposolda: I've added the corresponding adjustments to scope mapping to this PR. Currently some model tests are failing, but these are also failing on master. I can rebase when master is fixed.

**Review by pedroigor — DISMISSED** [review]
@mposolda I'm approving this one but I would like to hear from you what you think. Mainly because this also touches scope mapping and introduces a new behavior.

**Comment by mposolda** [comment]
@danielFesenmeyer Thanks for the PR! Sorry for the late review. The changes look good to me and I don't have any more comments. Just one thing: Can you please rebase to latest master? This will allow to run admin console tests in our internal testing pipeline as in theory there can be some regression in them. If you have a chance to doublecheck admin console tests in your environment, it will be beneficial as well. Thanks!

**Comment by danielFesenmeyer** [comment]
@mposolda Thanks for the review. I'll rebase to master and also doublecheck admin console tests - will let you know when finished.

**Comment by mposolda** [comment]
> @mposolda Thanks for the review. I'll rebase to master and also doublecheck admin console tests - will let you know when finished.

Thanks!

**Comment by danielFesenmeyer** [comment]
@mposolda: Rebased to a commit from yesterday. Cross-DC tests have failures, but these are the same as on master ("org.jboss.arquillian.container.spi.client.container.LifecycleException: Infinispan server startup timed out").
Admin console tests all worked fine (with latest chrome in headless mode).

**Comment by mposolda** [comment]
@danielFesenmeyer Thanks, trying pipeline build in our CI environment and will merge once it pass.

**Review by mposolda — APPROVED** [review]
@danielFesenmeyer Thanks! I've checked in our pipeline that cross-dc, cluster and admin console tests are fine. Just for the reference:
https://keycloak-jenkins.com/job/universal-test-pipeline-server/788/
https://keycloak-jenkins.com/job/universal-test-pipeline-server/791/

**Comment by danielFesenmeyer** [comment]
Thanks to you maintainers for reviewing and merging :)

**Comment by wwwy3y3** [comment]
hi @mposolda @pedroigor @stianst

Thanks for the review on this PR, which is really helpful in our use case.

Just want to ask if there's an estimated time of the `16.0.0` release  (I assume it would be the next release version, which includes this patch ?) since it's been 3 months since the last release.

Thanks for the hard work on Keycloak :)

**Pull request merged**
